File: scrape_4d.py
#!/usr/bin/env python3
"""
Singapore Pools 4D Results Scraper (Python version)

Usage: python3 scrape_4d.py [startDrawNo] [endDrawNo]
Example: python3 scrape_4d.py 5000 5429
"""

import urllib.request
import base64
import re
import json
import os
import sys
import time

BASE_URL = 'https://www.singaporepools.com.sg/en/product/Pages/4D_results.aspx?sppl='

def encode_draw_number(draw_no):
    return base64.b64encode(f'DrawNumber={draw_no}'.encode()).decode()

def fetch_page(url):
    req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
    with urllib.request.urlopen(req, timeout=15) as resp:
        return resp.read().decode('utf-8', errors='replace')

def parse_result(html, draw_no):
    try:
        date_match = re.search(r'class="drawDate"[^>]*>([^<]+)', html)
        draw_date = date_match.group(1).strip() if date_match else None

        first_match = re.search(r'class="tdFirstPrize"[^>]*>([^<]+)', html)
        second_match = re.search(r'class="tdSecondPrize"[^>]*>([^<]+)', html)
        third_match = re.search(r'class="tdThirdPrize"[^>]*>([^<]+)', html)

        # Extract starters
        starter_section = re.search(r'tbodyStarterPrizes[\s\S]*?</tbody>', html)
        starters = []
        if starter_section:
            nums = re.findall(r'<td[^>]*>(\d{4})</td>', starter_section.group(0))
            starters = nums

        # Extract consolation
        consol_section = re.search(r'tbodyConsolationPrizes[\s\S]*?</tbody>', html)
        consolation = []
        if consol_section:
            nums = re.findall(r'<td[^>]*>(\d{4})</td>', consol_section.group(0))
            consolation = nums

        if not first_match:
            return None

        # Parse date to ISO format
        iso_date = None
        if draw_date:
            # Remove day prefix like "Sun, " or "Wed, "
            cleaned = re.sub(r'^(Sun|Mon|Tue|Wed|Thu|Fri|Sat),?\s*', '', draw_date)
            from datetime import datetime
            for fmt in ('%d %B %Y', '%d %b %Y', '%B %d, %Y'):
                try:
                    dt = datetime.strptime(cleaned, fmt)
                    iso_date = dt.strftime('%Y-%m-%d')
                    break
                except ValueError:
                    continue

        return {
            'drawNo': draw_no,
            'date': iso_date or draw_date,
            'first': first_match.group(1).strip(),
            'second': second_match.group(1).strip() if second_match else None,
            'third': third_match.group(1).strip() if third_match else None,
            'starters': starters,
            'consolation': consolation
        }
    except Exception as e:
        print(f'  Error parsing draw {draw_no}: {e}')
        return None

def scrape_range(start, end):
    results_file = os.path.join(os.path.dirname(os.path.abspath(__file__)), '4d_results.json')
    existing = []

    if os.path.exists(results_file):
        with open(results_file) as f:
            existing = json.load(f)
        print(f'Loaded {len(existing)} existing results')

    existing_draws = {r['drawNo'] for r in existing}
    new_results = []
    errors = 0

    for draw_no in range(end, start - 1, -1):
        if draw_no in existing_draws:
            continue

        encoded = encode_draw_number(draw_no)
        url = BASE_URL + encoded

        try:
            print(f'Fetching draw {draw_no}...', end=' ', flush=True)
            html = fetch_page(url)
            result = parse_result(html, draw_no)

            if result:
                new_results.append(result)
                print(f'{result["date"]} - 1st: {result["first"]}, 2nd: {result["second"]}, 3rd: {result["third"]}')
                errors = 0
            else:
                print('No data')
                errors += 1

            # Rate limit
            time.sleep(0.8)

        except Exception as e:
            print(f'Error: {e}')
            errors += 1
            time.sleep(2)

        # Stop if too many consecutive errors (likely past valid range)
        if errors >= 10:
            print(f'\n10 consecutive errors, stopping at draw {draw_no}')
            break

        # Save progress every 50 draws
        if len(new_results) % 50 == 0 and len(new_results) > 0:
            all_results = existing + new_results
            all_results.sort(key=lambda r: r['drawNo'], reverse=True)
            seen = set()
            unique = [r for r in all_results if not (r['drawNo'] in seen or seen.add(r['drawNo']))]
            with open(results_file, 'w') as f:
                json.dump(unique, f, indent=2)
            print(f'  [Checkpoint] Saved {len(unique)} draws')

    # Final save
    all_results = existing + new_results
    all_results.sort(key=lambda r: r['drawNo'], reverse=True)
    seen = set()
    unique = [r for r in all_results if not (r['drawNo'] in seen or seen.add(r['drawNo']))]
    with open(results_file, 'w') as f:
        json.dump(unique, f, indent=2)
    print(f'\nTotal: {len(unique)} draws saved')
    print(f'New draws added: {len(new_results)}')

if __name__ == '__main__':
    start_draw = int(sys.argv[1]) if len(sys.argv) > 1 else 5000
    end_draw = int(sys.argv[2]) if len(sys.argv) > 2 else 5429

    print(f'Scraping draws {start_draw} to {end_draw}...')
    print(f'URL pattern: {BASE_URL}<base64("DrawNumber=XXXX")>')
    print()
    scrape_range(start_draw, end_draw)

File: scrape_4d.js
/**
 * Singapore Pools 4D Results Scraper
 *
 * Usage: node scrape_4d.js [startDrawNo] [endDrawNo]
 * Example: node scrape_4d.js 5300 5430
 *
 * URL Pattern discovered:
 *   https://www.singaporepools.com.sg/en/product/Pages/4D_results.aspx?sppl={base64("DrawNumber=XXXX")}
 *
 * The draw list (dates from Apr 2023 to present) is at:
 *   https://www.singaporepools.com.sg/DataFileArchive/Lottery/Output/fourd_result_draw_list_en.html
 *
 * The latest 6 draws are always at:
 *   https://www.singaporepools.com.sg/DataFileArchive/Lottery/Output/fourd_result_top_draws_en.html
 */

const https = require('https');
const fs = require('fs');

const BASE_URL = 'https://www.singaporepools.com.sg/en/product/Pages/4D_results.aspx?sppl=';

function encodeDrawNumber(drawNo) {
  return Buffer.from(`DrawNumber=${drawNo}`).toString('base64');
}

function fetchPage(url) {
  return new Promise((resolve, reject) => {
    https.get(url, { headers: { 'User-Agent': 'Mozilla/5.0' } }, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => resolve(data));
      res.on('error', reject);
    }).on('error', reject);
  });
}

function parseResult(html, drawNo) {
  try {
    // Extract draw date
    const dateMatch = html.match(/class="drawDate"[^>]*>([^<]+)/);
    const drawDate = dateMatch ? dateMatch[1].trim() : null;

    // Extract prizes
    const firstMatch = html.match(/class="tdFirstPrize"[^>]*>([^<]+)/);
    const secondMatch = html.match(/class="tdSecondPrize"[^>]*>([^<]+)/);
    const thirdMatch = html.match(/class="tdThirdPrize"[^>]*>([^<]+)/);

    // Extract starters
    const starterSection = html.match(/tbodyStarterPrizes[\s\S]*?<\/tbody>/);
    const starters = [];
    if (starterSection) {
      const nums = starterSection[0].match(/<td[^>]*>(\d{4})<\/td>/g);
      if (nums) nums.forEach(n => {
        const m = n.match(/>(\d{4})</);
        if (m) starters.push(m[1]);
      });
    }

    // Extract consolation
    const consolSection = html.match(/tbodyConsolationPrizes[\s\S]*?<\/tbody>/);
    const consolation = [];
    if (consolSection) {
      const nums = consolSection[0].match(/<td[^>]*>(\d{4})<\/td>/g);
      if (nums) nums.forEach(n => {
        const m = n.match(/>(\d{4})</);
        if (m) consolation.push(m[1]);
      });
    }

    if (!firstMatch) return null;

    // Parse date to ISO format
    let isoDate = null;
    if (drawDate) {
      const d = new Date(drawDate.replace(/^(Sun|Mon|Tue|Wed|Thu|Fri|Sat),?\s*/, ''));
      if (!isNaN(d.getTime())) {
        isoDate = d.toISOString().split('T')[0];
      }
    }

    return {
      drawNo,
      date: isoDate || drawDate,
      first: firstMatch[1].trim(),
      second: secondMatch ? secondMatch[1].trim() : null,
      third: thirdMatch ? thirdMatch[1].trim() : null,
      starters,
      consolation
    };
  } catch (e) {
    console.error(`Error parsing draw ${drawNo}:`, e.message);
    return null;
  }
}

async function scrapeRange(start, end) {
  const results = [];
  const existingFile = './4d_results.json';
  let existing = [];

  if (fs.existsSync(existingFile)) {
    existing = JSON.parse(fs.readFileSync(existingFile, 'utf8'));
    console.log(`Loaded ${existing.length} existing results`);
  }

  const existingDraws = new Set(existing.map(r => r.drawNo));

  for (let drawNo = end; drawNo >= start; drawNo--) {
    if (existingDraws.has(drawNo)) {
      console.log(`Draw ${drawNo} already exists, skipping`);
      continue;
    }

    const encoded = encodeDrawNumber(drawNo);
    const url = BASE_URL + encoded;

    try {
      console.log(`Fetching draw ${drawNo}...`);
      const html = await fetchPage(url);
      const result = parseResult(html, drawNo);

      if (result) {
        results.push(result);
        console.log(`  Got: ${result.date} - 1st: ${result.first}, 2nd: ${result.second}, 3rd: ${result.third}`);
      } else {
        console.log(`  No data found for draw ${drawNo}`);
      }

      // Rate limit: wait 1 second between requests
      await new Promise(r => setTimeout(r, 1000));
    } catch (e) {
      console.error(`  Error fetching draw ${drawNo}:`, e.message);
    }
  }

  // Merge with existing and sort by draw number descending
  const allResults = [...existing, ...results];
  allResults.sort((a, b) => b.drawNo - a.drawNo);

  // Remove duplicates
  const seen = new Set();
  const unique = allResults.filter(r => {
    if (seen.has(r.drawNo)) return false;
    seen.add(r.drawNo);
    return true;
  });

  fs.writeFileSync(existingFile, JSON.stringify(unique, null, 2));
  console.log(`\nTotal: ${unique.length} draws saved to ${existingFile}`);
  console.log(`New draws added: ${results.length}`);
}

// Main
const args = process.argv.slice(2);
const startDraw = parseInt(args[0]) || 5300;
const endDraw = parseInt(args[1]) || 5429;

console.log(`Scraping draws ${startDraw} to ${endDraw}...`);
console.log(`URL pattern: ${BASE_URL}<base64("DrawNumber=XXXX")>`);
console.log('');

scrapeRange(startDraw, endDraw);
